api/client/src/main/java/org/projectnessie/client/auth/oauth2/ImpersonationFlow.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2024 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.client.auth.oauth2;
+
+import java.net.URI;
+import java.util.Optional;
+import org.projectnessie.client.http.HttpAuthentication;
+
+/**
+ * A specialized {@link TokenExchangeFlow} that is performed after the initial token fetch flow, in
+ * order to obtain a more fine-grained token through impersonation (or delegation).
+ */
+class ImpersonationFlow extends TokenExchangeFlow {
+
+  private final ImpersonationConfig impersonationConfig;
+
+  ImpersonationFlow(OAuth2ClientConfig config) {
+    super(config);
+    impersonationConfig = config.getImpersonationConfig();
+  }
+
+  @Override
+  public Tokens fetchNewTokens(Tokens currentTokens) {
+    Tokens newTokens = super.fetchNewTokens(currentTokens);
+    return Tokens.of(newTokens.getAccessToken(), currentTokens.getRefreshToken());
```

Comment:
Here we keep: the exchanged access token + the _original_ refresh token (discarding any refresh token sent by the impersonation IdP, if any). I believe this is the correct behavior in all cases.

As for expiration: for all purposes, the validity of the exchanged token is the authoritative expiration time. The client will schedule a refresh at `min ( exchanged_token.expires_at, refresh_token == null ? NEVER : refresh_token.expires_at )`.